### PR TITLE
WeakChordalityInspector added

### PR DESCRIPTION
**Pull request by Toptachamann**
WeakChordalityInspector implemented. The implementation is based on this [article](http://www.ii.uib.no/~skeide/rec_wcg.pdf). The running time of the algorithm in O(|E|^2) and space complexity also O(|E|^2).

### CODE REVIEW

**Comment by jkinable** [comment]
Looks very good. Certainly a nice addition! 

It is pity that, as far as I can tell, the algorithm for weak chordality as presented in the thesis "Recognizing weakly chordal graphs" by Lars Severin Skeide (2002) never got published in a journal paper. Four years after the appearance of this thesis, the original 'inventor' of weak chordality published an O(|E|^2) algorithm ("Improved Algorithms for Weakly Chordal Graphs, 2006" https://webdocs.cs.ualberta.ca/~hayward/papers/iawcg.pdf), but this paper doesn't refer to Lars' thesis either. I wonder what's going on here?

A few more requests:

1. add a method GraphTests.isWeaklyChordal(.)
2. A graph is _not_ weakly chordal if it contains a hole or anti-hole. Similarly to the chordality inspector, the presence of a hole/anti-hole certifies that the graph is indeed not weakly chordal. Would it be possible to add a method which returns a hole/anti-hole in case the graph (after inspection) is not weakly chordal?

For (2), the detection of a hole/anti hole, the following references may be of help:

- Nikolopoulos, S.D. and Palios, L. Detecting holes and antiholes in graphs Algorithmica, 2007 Vol. 47, number 2, pages 119–138 http://www.cs.uoi.gr/~stavros/C-Papers/C-2004-SODA.pdf
- https://webdocs.cs.ualberta.ca/~hayward/papers/iawcg.pdf

**Comment by jkinable** [comment]
Following this PR, I've updated our TODO page:
https://github.com/jgrapht/jgrapht/wiki/TODOs
The only thing missing is a StrongChordalityInspector. That's obviously for a different PR :)

**Comment by Toptachamann** [comment]
Thanks for the class description, it is really better.

Regarding hole/anti-hole detection, in the paper by Hayward et al. the method to detect hole/anti-hole is based on the algorithm for recognition of weakly chordal graphs and it has O(n^2)  running time complexity as opposed to the O(m^2) of the algorithm by Nikolopoulos et al. 
At this point, I have a working version of the recognition part of the algorithm, so I'll figure out how to add hole/anti-hole detection.

The recognition part of the algorithm involves modifying the input graph, finding a neighborhood of the set of vertices, taking a subgraph of the graph (by the way, this is the reason this algorithm seems to me to be slower in practice than the version this PR is based on). In this PR I needed to find a neighborhood of an edge, for example. There is a method Graphs.neighborListOf(). The question is would it make sense to add analogous methods, which return sets of vertices?

I've also didn't come up with a clean way to copy the graph. The same situation with taking a subgraph. One way to do it is to use reflection, but it doesn't seem to be a good method. Should I use some other way to do it or have I missed something and this functionality is already implemented?

**Comment by jkinable** [comment]
Wrt to the definition of a 'Weakly chordal graph', I'm afraid there's still something unclear. I posted this as a question on math exchange:
https://math.stackexchange.com/questions/2722447/definition-of-weakly-chordal-graphs-and-difference-to-berge-graph/2722485#2722485
Perhaps you know the answer?

>Regarding hole/anti-hole detection, in the paper by Hayward et al. the method to detect hole/anti-hole is based on the algorithm for recognition of weakly chordal graphs and it has O(n^2) running time complexity as opposed to the O(m^2) of the algorithm by Nikolopoulos et al.
>At this point, I have a working version of the recognition part of the algorithm by Hayward et al, so I'll figure out how to add hole/anti-hole detection.

I'm not sure I understand you? The Hayward method with O(n^2) seems better than the Nikolopoulos  with O(m^2). It seems you are implementing the Hayward method, so that's great right?
Related to that: you already have code that is able to check whether the graph is weakly chordal. Isn't there some simple extension that gives you the desired proof (hole/anti-hole) if your graph isn't weakly chordal?

> I've also didn't come up with a clean way to copy the graph. The same situation with taking a subgraph. One way to do it is to use reflection, but it doesn't seem to be a good method. Should I use some other way to do it or have I missed something and this functionality is already implemented?

Copying graphs is unfortunately something we cannot do. This is really annoying. There are however some workarounds:

- if the vertex/edge type is irrelevant, you can make your own 'fake' copy by creating a new PseudoGraph with the same structure, perhaps by adding all vertices from the original graph and new DefaultEdges.
- you can use the SubGraph class to mask/hide edges/vertices
- you can use a technique similar to the one used for the implementation of the JohnsonShortestPaths class. This class requires as input a vertexfactory which tbh is not nice.

**Comment by Toptachamann** [comment]
Clearly, the question is how we define a hole.  I'll reference the papers I saw the definition of the hole in.
"An induced chordless cycle on five or more vertices in a graph is called a hole" [1]
"A hole is an induced cycle with five or more vertices and an antihole is the complement of a hole. A graph is weakly chordal (also called weakly triangulated) if it contains no holes and no antiholes" [2]

According to the definition from [this page](http://mathworld.wolfram.com/PerfectGraph.html), chordless cycles of 4 vertices can also be called holes. But the definition seems to be simply not precise enough. In my opinion, the definition of a hole as a chordless cycle on 5 or more vertices is more appropriate in the context of the weakly chordal graph. At least, it is used in both articles describing algorithms for recognition of weakly chordal graphs and it makes sense to me to use it in the documentation of the implementation. 

Regarding the hole detection. In the case of chordal graphs, it wasn't a big problem to figure out how to find a chordless cycle of 4 or more vertices using order computed by some iterator. Actually, now I can't say for sure whether it will be possible to do with the algorithm by Lars Skeide. If it will be possible to use current implementation to do, it will be the choice. For instance, in [2] the algorithm for weakly chordal graph recognition is mentioned, which can't be modified to detect a hole. So I need to read some literature and think more about it. 

[1] Lars Skeide. "Recognizing weakly chordal graphs"
[2] Hayward et al. "Improved Algorithms for Weakly Chordal Graphs"

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/WeakChordalityInspector.java [hunk 1]
```diff
@@ -0,0 +1,443 @@
+/*
+ * (C) Copyright 2018, by Timofey Chudakov and Contributors.
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/WeakChordalityInspector.java]: Before I forget: please keep the copyright header format identical. E.g. your first line should read:
 * (C) Copyright 2018-2018, by Timofey Chudakov and Contributors.

Headers are updated automatically; Deviation of the standard header messes this up.

**Comment by Toptachamann** [comment]
So, it turned out it is possible to modify the algorithm so that it detects holes or anti-hole. 

We find the first minimal separator **s** that is contained in the neighborhood some edge  **e** on the vertices **source** and **target**, which contains a coconnected component **c**, which has two vertices **sourceInSeparator** and **targetInSeprator**. **sourceInSeparator** is adjacent only to **source** and **targetInSeparator** is adjacent only to **target**. In the case, when **sourceInSeparator** isn't adjacent to **targetInSeparator** we find a path from **sourceInSeparator** to the **targetInSeparator** in the remaining part of the graph. Each vertex on this cycle isn't adjacent to the **source** and **target**. This condition is used to find the hole correctly. Such path exists, it follows from the way we computed minimal separator **s**: we used dfs from some vertex in the remaining graph, that is why the component that is separated by the separator **s** is connected and every vertex in the separator is connected to some vertex in this component. Every vertex in the separated component isn't adjacent to the vertices **source** and **target** by definition.  

In the case **sourceInSeparator** and **targetInSeparator** are adjacent we have detected an anti-hole. The problem here is these vertices aren't consecutive vertices of the anti-hole. In fact, (**source** and **targetInSeparator**) and (**target** and **sourceInSeparator**) lay consecutively on the anti-hole we detected. So I took the complement of the graph, computed minimal separators of the edge **e'** = (**source**, **targetInSeparator**) and repeated the process of finding a hole in the complement of the graph. This way may be not the most optimal algorithm in the case of sparse graphs, but it results in relatively simple code. 

Other changes:
* Created SeparatorFinder. Essentially, the functionality of finding a separator doesn't depend only on the WeakChordalityInspector and therefore could be decoupled. This class can be further expanded so that is finds global minimal separators. 
* Added two helper methos in the class Graphs.

**Comment by jkinable** [comment]
I've got the issues wrt the definition of weak chordality finally worked out. 

[Hayward, 1985] defines weakly chordal graphs, also known as weakly triangulated graphs as follows:

> we define a graph as weakly triangulated if neither it nor its complement contains a chordless cycle with five or more vertices
Hayward, Ryan B., Weakly triangulated graphs, J. Comb. Theory, Ser. B 39, 200-208 (1985). ZBL0551.05055.

Douglas West, Introduction to graph theory maintains the same definition.

http://www.graphclasses.org/classes/gc_14.html defines it as:
A graph is weakly chordal if it is (anti-hole,hole)-free.
It appears that the definition of a hole is not well defined. Some sources define it as a chordless cycle of lenght 4, 5 or 6. However, graphclasses.org defines a hole as a chordless cycle of length 5: http://graphclasses.org/smallgraphs.html#holes

So in conclusion, their definition of a weakly chordal graph is the same is ours. Nevertheless, I would prefer to be very specific about this, and use the definition of [Hayward, 1985] as defined above.

**Comment by Toptachamann** [comment]
So, if I understand you right, in the header Javadoc of the WeakChordalityInspector we don't use the term "hole" and state explicitly in the definition of weakly chordal graphs, that neither they nor their complements don't contain chordless cycles with 5 or more vertices (which makes sense to me in this situation).

File: jgrapht-core/src/main/java/org/jgrapht/GraphTests.java [hunk 2]
```diff
@@ -496,6 +497,32 @@
         return new ChordalityInspector<>(graph).isChordal();
     }
 
+    /**
+     * Checks whether a graph is <a href="http://www.graphclasses.org/classes/gc_14.html">weakly chordal</a>.
+     * <p>
+     * The following definitions are equivalent:
+     * <ol>
+     * <li> A graph is weakly chordal if it is (<a href="http://mathworld.wolfram.com/GraphAntihole.html">
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/GraphTests.java]: Let's make the description identical to the one found in WeakChordalityInspector

File: jgrapht-core/src/main/java/org/jgrapht/alg/cycle/WeakChordalityInspector.java [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+ * <a href="http://www.ii.uib.no/~skeide/rec_wcg.pdf"><i>Recognizing weakly chordal graphs</i></a>.
+ * Candidate Scientist Thesis in Informatics. Department of Informatics, University of Bergen, Norway.
+ * The terminology used in this implementation is consistent with the one used in this thesis.
+ * <p>
+ * Both the runtime complexity and space complexity of the algorithm implemented in this class is $\mathcal{O}(|E|^2)$.<br>
+ * The inspected {@code graph} is specified at the construction time and cannot be modified.
+ * When the graph is modified externally, the behavior of the {@code WeakChordalityInspector} is undefined.
+ * <p>
+ * In the case the inspected graph in not weakly chordal, this inspector provides a certificate in the form
+ * of some hole or anti-hole. The running time of finding a hole is $\mathcal{O}(|V| + |E|)$ and of
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/cycle/WeakChordalityInspector.java]: To avoid confusion, let's link to the hole/anti-hole definitions on the graphclasses website, instead of to the wolfram website:
hole: http://graphclasses.org/smallgraphs.html#holes
anti-hole: http://graphclasses.org/smallgraphs.html#antiholes
